satpy/readers/gms/gms5_vissr_navigation.py

```diff
@@ -0,0 +1,932 @@
+        observation_time,
+        attitude_prediction.prediction_times,
+        attitude_prediction.attitude.angle_between_sat_spin_and_z_axis,
+    )
+    angle_between_sat_spin_and_yz_plane = interpolate_angles(
+        observation_time,
+        attitude_prediction.prediction_times,
+        attitude_prediction.attitude.angle_between_sat_spin_and_yz_plane,
+    )
+    return Attitude(
+        angle_between_earth_and_sun,
+        angle_between_sat_spin_and_z_axis,
+        angle_between_sat_spin_and_yz_plane,
+    )
+
+
+@numba.njit
+def interpolate_continuous(x, x_sample, y_sample):
+    """Linear interpolation of continuous quantities.
+
+    Numpy equivalent would be np.interp(..., left=np.nan, right=np.nan), but
+    numba currently doesn't support those keyword arguments.
+    """
+    try:
+        return _interpolate(x, x_sample, y_sample)
+    except Exception:
+        # Numba cannot distinguish exception types
+        return np.nan
+
+
+@numba.njit
+def _interpolate(x, x_sample, y_sample):
+    i = _find_enclosing_index(x, x_sample)
+    offset = y_sample[i]
+    x_diff = x_sample[i + 1] - x_sample[i]
+    y_diff = y_sample[i + 1] - y_sample[i]
+    slope = y_diff / x_diff
+    dist = x - x_sample[i]
+    return offset + slope * dist
+
+
+@numba.njit
+def _find_enclosing_index(x, x_sample):
+    """Find where x_sample encloses x."""
+    for i in range(len(x_sample) - 1):
+        if x_sample[i] <= x < x_sample[i + 1]:
+            return i
+    raise Exception("x not enclosed by x_sample")
+
+
+@numba.njit
+def interpolate_angles(x, x_sample, y_sample):
+    """Linear interpolation of angles.
+
+    Requires 2-pi periodicity to be unwrapped before (for
+    performance reasons). Interpolated angles are wrapped
+    back to [-pi, pi] to restore periodicity.
+    """
+    return _wrap_2pi(interpolate_continuous(x, x_sample, y_sample))
+
+
+@numba.njit
+def _wrap_2pi(values):
+    """Wrap values to interval [-pi, pi].
+
+    Source: https://stackoverflow.com/a/15927914/5703449
+    """
+    return (values + np.pi) % (2 * np.pi) - np.pi
+
+
+@numba.njit
+def interpolate_nearest(x, x_sample, y_sample):
+    """Nearest neighbour interpolation."""
+    try:
+        return _interpolate_nearest(x, x_sample, y_sample)
+    except Exception:
+        return np.nan * np.ones_like(y_sample[0])
+
+
+@numba.njit
+def _interpolate_nearest(x, x_sample, y_sample):
+    i = _find_enclosing_index(x, x_sample)
+    return y_sample[i]
+
+
+@numba.njit
+def matrix_vector(m, v):
+    """Multiply (3,3)-matrix and Vector3D."""
+    x = m[0, 0] * v.x + m[0, 1] * v.y + m[0, 2] * v.z
+    y = m[1, 0] * v.x + m[1, 1] * v.y + m[1, 2] * v.z
+    z = m[2, 0] * v.x + m[2, 1] * v.y + m[2, 2] * v.z
+    return Vector3D(x, y, z)
+
+
+@numba.njit
+def cross_product(a, b):
+    """Compute vector product a x b."""
+    return Vector3D(
+        x=a.y * b.z - a.z * b.y,
+        y=a.z * b.x - a.x * b.z,
+        z=a.x * b.y - a.y * b.x
+    )
```

Comment:
Both these functions could be replaced with equivalent numpy functions (implemented in numba): `np.dot` and `np.cross`. Question is why did you feel it was necessary to use named tuples here and an own implementation?

Reply:
I used the numpy functions in the beginning, but that was quite slow. I think the reason is that new numpy arrays (3d vectors) are allocated in each iteration of the main loop over all pixels. And although they are very small and always have 3 elements, numba apparently can't optimize that at compile time. With namedtuples (which are translated to C structs) numba is doing a better job.

An alternative solution is to allocate those arrays outside the main loop and pass them all the way down to the respective method. But I found that not very clean.

Reply:
Ok, I understand, let's leave it like this